libs/community/langchain_community/chat_models/baichuan.py

```diff
@@ -222,14 +248,96 @@ def _stream(
                     run_manager.on_llm_new_token(chunk.content, chunk=cg_chunk)
                 yield cg_chunk
 
+    async def _agenerate(
+        self,
+        messages: List[BaseMessage],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
+        stream: Optional[bool] = None,
+        **kwargs: Any,
+    ) -> ChatResult:
+        should_stream = stream if stream is not None else self.streaming
+        if should_stream:
+            stream_iter = self._astream(
+                messages, stop=stop, run_manager=run_manager, **kwargs
+            )
+            return await agenerate_from_stream(stream_iter)
+
+        headers = self._create_headers_parameters(**kwargs)
+        payload = self._create_payload_parameters(messages, **kwargs)
+
+        import httpx
+
+        async with httpx.AsyncClient(
+            headers=headers, timeout=self.request_timeout
+        ) as client:
+            response = await client.post(self.baichuan_api_base, json=payload)
+            response.raise_for_status()
+        return self._create_chat_result(response.json())
+
+    async def _astream(
+        self,
+        messages: List[BaseMessage],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> AsyncIterator[ChatGenerationChunk]:
+        headers = self._create_headers_parameters(**kwargs)
+        payload = self._create_payload_parameters(messages, stream=True, **kwargs)
+        import httpx
+
+        async with httpx.AsyncClient(
+            headers=headers, timeout=self.request_timeout
+        ) as client:
+            async with aconnect_httpx_sse(
+                client, "POST", self.baichuan_api_base, json=payload
+            ) as event_source:
+                async for sse in event_source.aiter_sse():
+                    chunk = json.loads(sse.data)
+                    if len(chunk["choices"]) == 0:
+                        continue
+                    choice = chunk["choices"][0]
+                    chunk = _convert_delta_to_message_chunk(
+                        choice["delta"], AIMessageChunk
+                    )
+                    finish_reason = choice.get("finish_reason", None)
+
+                    generation_info = (
+                        {"finish_reason": finish_reason}
+                        if finish_reason is not None
+                        else None
+                    )
+                    chunk = ChatGenerationChunk(
+                        message=chunk, generation_info=generation_info
+                    )
+                    yield chunk
+                    if run_manager:
+                        await run_manager.on_llm_new_token(chunk.text, chunk=chunk)
```

Comment:
This callback should be invoked prior to yielding the chunk.

https://github.com/langchain-ai/langchain/issues/16913

Reply:
done